driver-core/src/main/com/mongodb/internal/connection/DefaultAuthenticator.java

```diff
@@ -51,8 +52,10 @@ void authenticate(final InternalConnection connection, final ConnectionDescripti
                     .authenticate(connection, connectionDescription);
         } else {
             try {
-                BsonDocument isMasterResult = executeCommand("admin", createIsMasterCommand(), connection);
-                getAuthenticatorFromIsMasterResult(isMasterResult, connectionDescription)
+                Authenticator authenticator = getAuthenticatorForSpeculativeAuthentication();
```

Comment:
This class should override `createSpeculativeAuthenticateCommand`.  Otherwise there won't be speculation for the common case where the auth mechanism is not specified (and DefaultAuthenticator is used).

Reply:
By the time DefaultAuthenticator.authenticate is called, I think you've already lost your opportunity to do speculative authentication.  I proved this by stepping through the code using a connection string like `mongodb://user:pass&@localhost` and observing that no speculative authentication occurs.

Reply:
Looks logically correct now, so resolving this thread.